### ISSUE TITLE
🐛 Bug Report: UnifiedThemeOptions has no way to override some Backstage component styles

### ISSUE THREAD

**Issue by joshjung** [backstage/backstage#26754]
### 📜 Description

Using the V4 theming, you could provide any overrides for any components. For example, you could provide the following override:

```typescript
const highlightOverride = {
  BackstageHighlightedSearchResultText: {
    highlight: {
      color: 'inherit',
      backgroundColor: 'inherit',
      fontWeight: 'bold',
      textDecoration: 'underline',
    },
  },
};
```

See: https://backstage.io/docs/features/search/how-to-guides/#how-to-customize-search-results-highlighting-styling

But now, the `UnifiedThemeOptions` object provides no type-safe way to override a component like this, because the typing is strictly tied to the V5 MUI components:

```
import {...ThemeOptions as ThemeOptionsV5...} from '@mui/material/styles';

export interface UnifiedThemeOptions {
  palette: PaletteOptionsV4 & PaletteOptionsV5;
  defaultPageTheme?: string;
  pageTheme?: Record<string, PageTheme>;
  fontFamily?: string;
  htmlFontSize?: number;
  components?: ThemeOptionsV5['components'];
  typography?: BackstageTypography;
}
```

Using the new `createUnifiedTheme` recommended for providing theming that works for V4 and V5 there is no type-safe way to follow the docs to override styles for non-MUI provided components.

### 👍 Expected behavior

Should be able to follow the docs at least which instruct on how to style the `BackstageHighlightedSearchResultText`.

### 👎 Actual Behavior with Screenshots

See above.

### 👟 Reproduction steps

See above.

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by awanlin** [comment]
Hi @joshjung, looks like the search docs you linked need to be corrected. This works: https://backstage.io/docs/getting-started/app-custom-theme#overriding-backstage-and-material-ui-components-styles as it's used heavily in the `aperture` theme - https://github.com/backstage/demo/blob/9083e49114dadc4cf5456fae5fce566bdc3fb5fc/packages/app/src/theme/aperture.ts#L118 - used in the Demo site.

**Comment by joshjung** [comment]
Looks like this is not just the docs, though, because I have to specify a @tsignore:

![Screenshot 2024-09-17 at 18 04 38](https://github.com/user-attachments/assets/fae9607c-600e-423c-b9dd-b8730348ea4a)

**Comment by joshjung** [comment]
Okay, it took me a while to figure out what was going on, but I think I know what I need to know to make a PR. Will get one ready...

**Comment by joshjung** [comment]
Okay, nevermind, @Rugvip I think I'm going to need your help here. I see the core problem.

It's clear to me now that you created the `OverrideComponentNameToClassKeys` interface as a way to inject new keys into the `'@mui/material/styles'` `interface Components` so that TypeScript can see which keys are available. Then that interface is imported into the new UnifiedTheme object `components` field type.

Right now only the `plugins/core-components` plugin provides its components to that new interface, which means every single other plugin that has UX components does not update the `'@mui/material/styles'` `interface Components` with its types.

The big question is, how can every single plugin in use by a Backstage app provide more keys to the `'@mui/material/styles'` `interface Components` in an organized way? This would need to be done for dozens of plugins...

There is still a definite bug. That bug is in this code here:

```
/**
 * Transform Material UI v5 component themes into a v4 theme props and overrides.
 *
 * @public
 */
export function transformV5ComponentThemesToV4(
  theme: Theme,
  components: ThemeOptions['components'] = {},
): { overrides: Overrides; props: ComponentsProps } {
  ...
  for (const name of Object.keys(components)) {
    const component = components[name as keyof typeof components];
    if (!component) {
      continue;
    }
    ...
```

Basically, it appears that even if I use @tsignore and try to provide a component that does not exist in the type, that `keyof typeof` throws out the css style definition, which seems a little restrictive. If I am reading this right.

**Comment by joshjung** [comment]
Got it, this works:

```
export const myLightTheme : UnifiedTheme = createUnifiedTheme({
  ...createBaseThemeOptions({
    palette: palettes.light,
  }),
  defaultPageTheme: 'home',
  components: {
    /** @ts-ignore */
    BackstageHighlightedSearchResultText: {
      styleOverrides: {
        highlight: {
          color: 'inherit',
          backgroundColor: 'red'
        }
      }
    }
  }
});
```

I will create a PR to update the documentation.

**Comment by awanlin** [comment]
Should be like this:

```ts
BackstageHighlightedSearchResultText: {
      styleOverrides: {
        highlight: {
          color: 'inherit',
          backgroundColor: 'red',
        },
      },
    },
```

Not sure about the types issue though. Most plugins that have overrides have a `overridableComponents.ts` file with them. Also see this issue: https://github.com/backstage/backstage/issues/25748

**Comment by awanlin** [comment]
Ah, cool, posted at the same time 👍

**Comment by joshjung** [comment]
I'm not sure why it doesn't see the pull request that references this issue: https://github.com/backstage/backstage/pull/26757
